modules/renter/proto/merkleroots.go

```diff
@@ -0,0 +1,343 @@
+package proto
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/crypto"
+	"github.com/NebulousLabs/errors"
+)
+
+// merkleRootsCacheHeight is the height of the subTrees in cachedSubTrees. A
+// height of 7 means that 128 sector roots are covered by a single cached
+// subTree.
+const merkleRootsCacheHeight = 7
+
+type (
+	// merkleRoots is a helper struct that makes it easier to add/insert/remove
+	// merkleRoots within a SafeContract.
+	merkleRoots struct {
+		// cachedSubTrees are cached trees that can be used to more efficiently
+		// compute the merkle root of a contract. The cached trees are not
+		// persisted and are computed after startup.
+		cachedSubTrees []*cachedSubTree
+		// uncachedRoots contains the sector roots that are not part of a
+		// cached subTree. The uncachedRoots slice should never get longer than
+		// 2^merkleRootsCacheHeight since that would simply result in a new
+		// cached subTree in cachedSubTrees.
+		uncachedRoots []crypto.Hash
+
+		// file is the file of the safe contract that contains the root.
+		file *os.File
+		// numMerkleRoots is the number of merkle roots in file.
+		numMerkleRoots int
+	}
+
+	cachedSubTree struct {
+		height int
+		sum    crypto.Hash
+	}
+)
+
+// loadExistingMerkleRoots reads creates a merkleRoots object from existing
+// merkle roots.
+func loadExistingMerkleRoots(file *os.File) (mr *merkleRoots, err error) {
+	mr = &merkleRoots{
+		file: file,
+	}
+	// Get the number of roots stored in the file.
+	mr.numMerkleRoots, err = mr.lenFromFile()
+	if err != nil {
+		return
+	}
+	// Seek to the first root's offset.
+	if _, err = file.Seek(rootIndexToOffset(0), io.SeekStart); err != nil {
+		return
+	}
+	// Read the roots from the file without reading all of them at once.
+	for i := 0; i < mr.numMerkleRoots; i++ {
+		var root crypto.Hash
+		if _, err = io.ReadFull(file, root[:]); err == io.EOF {
+			break
+		} else if err != nil {
+			return
+		}
+
+		// Append the root to the unachedRoots
+		mr.uncachedRoots = append(mr.uncachedRoots, root)
+
+		// If the uncachedRoots grew too large we add them to the cache.
+		if len(mr.uncachedRoots) == (1 << merkleRootsCacheHeight) {
+			st := newCachedSubTree(mr.uncachedRoots)
+			mr.cachedSubTrees = append(mr.cachedSubTrees, st)
+			mr.uncachedRoots = mr.uncachedRoots[:0]
+		}
+	}
+	return mr, nil
+}
+
+// newCachedSubTree creates a cachedSubTree from exactly
+// 2^merkleRootsCacheHeight roots.
+func newCachedSubTree(roots []crypto.Hash) *cachedSubTree {
+	// Sanity check the input length.
+	if len(roots) != (1 << merkleRootsCacheHeight) {
```

Comment:
So, height of 0 means it's the hash of a leaf, height of 1 means it's the root of 2 leaves, height of 2 means root of 4 leaves, etc?

I don't think that's clearly documented in the merkle tree cached functions, we should make sure the docstrings for all of the cached roots properly explain what counts as height 0 and 1 etc.